Fixes #27442 - deprecation mysql warning

lzap opened the pull request:
A notification that warns about MySQL in a non-intrusive way similarly that we did for deprecated settings about a year ago.

Edit: Based off: https://github.com/theforeman/foreman/pull/6038/files

lzap committed:
Fixes #27442 - deprecation mysql warning

lzap commented:
![obrazek](https://user-images.githubusercontent.com/49752/62041214-6ae7e200-b1fb-11e9-85e9-2bb67d48d87f.png)

This points to a non-existing blogpost: https://github.com/theforeman/theforeman.org/pull/1404

ares commented on config/application.rb:
```diff
@@ -343,6 +344,30 @@ def init_dynflow
+    def notify_deprecations
+      if ActiveRecord::Base.connection.adapter_name.downcase.starts_with?('mysql')
+        blueprint = NotificationBlueprint.find_by_name('feature_deprecation')
+        Foreman::Deprecation.deprecation_warning('1.24', "Support for MySQL is deprecated, migrate to PostgreSQL")
```
there was a discussion to postpone this by one version, is it 1.24 now then?

lzap replied:
Yeah, changing to 1.25.

ares reviewed (commented):
looks reasonable to me (haven't tested yet), confirmation from @tbrisker and @ekohl who were mostly active in the discourse thread would be great

given it works, I wouldn't mind merging, of there are no objections

ezr-ondrej commented on config/application.rb:
```diff
@@ -343,6 +344,30 @@ def init_dynflow
       Audit.send(:include, AuditSearch)
     end
+
+    def notify_deprecations
```
wouldn't it be better to have a sepparate `config/initializers/<i.e. depracate_mysql.rb>` file?

lzap replied:
Yes, we actually have an existing `deprecations.rb` initializer, so moving it there.

ezr-ondrej reviewed (commented):
The implementation looks quite good. I would move it to separate file, though...

tbrisker commented:
I'll be ok with pulling this in post branching, and would prefer to properly review the migration instructions and blog post not on the same day as branching. I think also 1.25 is more reasonable so users have enough time to prepare and for us to test the migration, and i'll comment to that effect on the discourse thread.

ohadlevy commented on config/application.rb:
```diff
@@ -343,6 +344,30 @@ def init_dynflow
+    def notify_deprecations
+      if ActiveRecord::Base.connection.adapter_name.downcase.starts_with?('mysql')
+        blueprint = NotificationBlueprint.find_by_name('feature_deprecation')
+        Foreman::Deprecation.deprecation_warning('1.24', "Support for MySQL is deprecated, migrate to PostgreSQL")
```
maybe consider adding "please migrate" to the text (or make it sound less like an order :)

lzap replied:
Good idea.

ares commented:
> I'll be ok with pulling this in post branching, and would prefer to properly review the migration instructions and blog post not on the same day as branching.

ack, I won't touch it :-)

lzap commented:
Incorporated all comments, rebased, fixed bug when blueprint was not present (database was not yet seeded). I am using "1.25/2.0" in the notification text, however I keep "1.25" in `Foreman::Deprecation.deprecation_warning` because we use that in our codebase and we will likely replace this once we do the final decision about 2.0.

ezr-ondrej reviewed (dismissed):
I hate rockets (`=>`) :rocket:
Otherwise LGTM :+1:

tbrisker commented:
Mysql tests are failing, i guess because of the deprecation warning - we need to add it to the whitelist

tbrisker commented on config/initializers/deprecations.rb:
```diff
@@ -1 +1,30 @@
+        Foreman::Deprecation.deprecation_warning('1.25', "Support for MySQL is deprecated, please migrate to PostgreSQL")
+        blueprint = NotificationBlueprint.find_by_name('feature_deprecation')
+        return unless blueprint
+        message = UINotifications::StringParser.new(blueprint.message, {feature: 'MySQL as a backend database', version: '1.25/2.0'}).to_s
```
let's keep this as 1.25 for now until we decide on the version number, we can always change it later.

lzap commented:
Removed 2.0, added to the whitelist, rebased.

ezr-ondrej commented on config/initializers/deprecations.rb:
```diff
@@ -1 +1,30 @@
+        Foreman::Deprecation.deprecation_warning('1.25', "Support for MySQL is deprecated, please migrate to PostgreSQL")
+        blueprint = NotificationBlueprint.find_by_name('feature_deprecation')
+        return unless blueprint
+        message = UINotifications::StringParser.new(blueprint.message, {feature: 'MySQL as a backend database', version: '1.25'}).to_s
```
Other notifications got its classes and define theirs `initialize`, if we want to have a generick deprecation class, wouldn't it great to have such class, which would encapsulate these?

lzap replied:
Sorry what? ;-) Can you provide an example code?

tbrisker commented on config/initializers/deprecations.rb:
```diff
@@ -1 +1,30 @@
+          :actions => {
+            :links => [
+              {
+                :href => 'https://theforeman.org/2019/08/dropping-support-for-mysql.html',
```
We'll need to update this once we get the post published.

tbrisker replied:
https://theforeman.org/2019/09/dropping-support-for-mysql.html

tbrisker reviewed (changes requested):
please update the link and rebase on latest develop, otherwise gtg

lzap commented:
Done.

tbrisker reviewed (dismissed):
Thanks @lzap ! just installed mysql to verify and this works well.

tbrisker commented on config/as_deprecation_whitelist.yaml:
```diff
@@ -14,3 +14,6 @@
     with non-attribute argument(s):'
+
+# https://projects.theforeman.org/issues/27035
+- message: 'Support for MySQL is deprecated, migrate to PostgreSQL'
```
Looks like this only works inside tests? mysql tests are failing to run with:
`/usr/local/rvm/gems/ruby-2.5.1@test_develop_pr_core-3/gems/activesupport-5.2.1/lib/active_support/lazy_load_hooks.rb:69:in block in execute_hook': DEPRECATION WARNING: You are using a deprecated behavior, it will be removed in version 1.25, Support for MySQL is deprecated, please migrate to PostgreSQL (called from <top (required)> at /home/jenkins/workspace/test_develop_pr_core/database/mysql/ruby/2.5/slave/fast/config/environment.rb:5) (ActiveSupport::DeprecationException)`

mmoll replied:
The full message is needed, try:

```
# https://projects.theforeman.org/issues/27035
- message: 'You are using a deprecated behavior, it will be removed in version 1.25,
    Support for MySQL is deprecated, migrate to PostgreSQL'
```

lzap commented:
Rebased.

tbrisker commented:
sadly looks like that doesn't work - it is still failing on mysql with `/usr/local/rvm/gems/ruby-2.5.1@test_develop_pr_core-2/gems/activesupport-5.2.1/lib/active_support/lazy_load_hooks.rb:69:in block in execute_hook': DEPRECATION WARNING: You are using a deprecated behavior, it will be removed in version 1.25, Support for MySQL is deprecated, please migrate to PostgreSQL (called from <top (required)> at /home/jenkins/workspace/test_develop_pr_core@2/database/mysql/ruby/2.5/slave/fast/config/environment.rb:5) (ActiveSupport::DeprecationException)` - i'm guesing the whitelist only works during the tests themselves and not the initializers

lzap commented:
Removed, should be good now.

lzap commented:
Unrelated - weird failure? Cannot see it.

tbrisker commented on config/initializers/deprecations.rb:
```diff
@@ -1 +1,29 @@
+    unless Foreman.in_rake?
+      if ActiveRecord::Base.connection.adapter_name.downcase.starts_with?('mysql')
+        blueprint = NotificationBlueprint.find_by_name('feature_deprecation')
+        return unless blueprint
```
there is no function - you are in a block, you can't return, you should use `break` instead:
```suggestion
        break unless blueprint
```

tbrisker commented on config/initializers/deprecations.rb:
```diff
@@ -1 +1,29 @@
+
+Foreman::Application.configure do |app|
+  config.after_initialize do
+    unless Foreman.in_rake?
```
```suggestion
    break if Foreman.in_rake?
```
will reduce indentation level :)

tbrisker commented on config/initializers/deprecations.rb:
```diff
@@ -1 +1,29 @@
+        blueprint = NotificationBlueprint.find_by_name('feature_deprecation')
+        return unless blueprint
+        message = UINotifications::StringParser.new(blueprint.message, {feature: 'MySQL as a backend database', version: '1.25'}).to_s
+        return if blueprint.notifications.where(message: message).any?
```
This should also be `break`

tbrisker commented on config/initializers/deprecations.rb:
```diff
@@ -1 +1,29 @@
+Foreman::Application.configure do |app|
+  config.after_initialize do
+    unless Foreman.in_rake?
+      if ActiveRecord::Base.connection.adapter_name.downcase.starts_with?('mysql')
```
can also reduce indent level:
```suggestion
      break unless ActiveRecord::Base.connection.adapter_name.downcase.starts_with?('mysql')
```

lzap commented:
All right rebased now hopefully for the last time.

tbrisker reviewed (approved):
Thanks @lzap ! test failure unrelated

tbrisker commented:
1.23 - 2c7aa4480
